strongbox-storage/strongbox-storage-core/src/main/java/org/carlspring/strongbox/io/ArtifactInputStream.java

```diff
@@ -152,6 +177,10 @@ public int read(byte[] bytes,
                 MessageDigest digest = (MessageDigest) entry.getValue();
                 digest.update(bytes, off, numberOfBytesRead);
             }
+            if(isInputStreamValid()){
+                this.extension = getFileExtension(bytes);
```

Comment:
The problem with this approach is that you have to have read the whole file in memory in order to pass the array of bytes. How much memory would you need to have, if you had 50 different sessions deploying artifacts each of around 200-300 MB (let's assume they're `war` files). This would be very memory inefficient, won't it?

Couldn't we just have some sort of `InputStream` wrapper, or something, that just uses the bytes that have already been ready (which we're already passing to the `digest.update((byte)...)` method? If we could implement something like this, then things would be significantly more efficient and lightweight.

What do you think?

Reply:
That's what I had been trying to for last 10 days, but the problem is that no matter what warpper you choose BufferedOutputStream/inputStream, the moment you try and read those bytes, inputStream becomes invalid for some of the web-core test like "testMetaDataMerge" or "testPackageCommonFlow" as the stream may not support mark and reset.

As far as trying to avoid memory buffer goes, whether we create a wrapper object like ByteOutputStream --> convert to bytes --> BufferedInoutStream for example, will still take up memory space when you are trying to convert to a parallel stream.

I have used this code before to read the stream twice but here in this project, I get message that some of the test based on some test resources, the files are not mark() supoorted.
`ByteArrayOutputStream baos = new ByteArrayOutputStream();
org.apache.commons.io.IOUtils.copy(in, baos);
byte[] bytes = baos.toByteArray();

// either
while (needToReadAgain) {
    ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
    yourReadMethodHere(bais);
}

// or
ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
while (needToReadAgain) {
    bais.reset();
    yourReadMethodHere(bais);
}`

So if we want to use a parallel stream then we will need to revisit the test cases in web-core domain. The sample test resource  files that are being used there have issues and do not support mark and reset.